File: apps/api/app/schemas/guardrails.py
from typing import Literal

from pydantic import Field

from app.schemas import ApiSchema

RuleType = Literal[
    "restricted_category",
    "forbidden_phrase",
    "escalation_trigger",
    "sensitive_topic",
    "response_constraint",
]
TargetCategory = Literal[
    "legal_judgment",
    "outcome_prediction",
    "definitive_benefit_decision",
    "unsupported_amount_confirmation",
    "unsupported_deadline_confirmation",
    "insufficient_evidence",
    "conflict_detected",
    "restricted_topic_detected",
    "repeated_user_dissatisfaction",
    "after_hours_routing",
    "welfare_eligibility_judgment",
    "legal_interpretation",
    "administrative_decision_prediction",
    "risky_civic_complaint_request",
    "cautious_wording",
    "warning_notice",
    "escalation_suggestion",
    "clarification_required",
]
MatchMode = Literal["keyword_any", "contains", "exact", "context_flag"]
ActionType = Literal[
    "restricted",
    "escalate",
    "ask_clarification",
    "fallback",
    "warn",
    "require_cautious_wording",
]
Severity = Literal["low", "medium", "high", "critical"]


class GuardrailRuleBase(ApiSchema):
    rule_type: RuleType
    target_category: TargetCategory | None = None
    match_mode: MatchMode = "keyword_any"
    match_value: str | None = Field(default=None, max_length=2000)
    action_type: ActionType
    severity: Severity = "medium"
    fallback_message: str | None = Field(default=None, max_length=2000)
    escalation_message: str | None = Field(default=None, max_length=2000)
    priority: int = Field(default=100, ge=1, le=1000)
    is_active: bool = True
    metadata_json: dict | None = None


class GuardrailRuleCreateRequest(GuardrailRuleBase):
    pass


class GuardrailRuleUpdateRequest(ApiSchema):
    target_category: TargetCategory | None = None
    match_mode: MatchMode | None = None
    match_value: str | None = Field(default=None, max_length=2000)
    action_type: ActionType | None = None
    severity: Severity | None = None
    fallback_message: str | None = Field(default=None, max_length=2000)
    escalation_message: str | None = Field(default=None, max_length=2000)
    priority: int | None = Field(default=None, ge=1, le=1000)
    is_active: bool | None = None
    metadata_json: dict | None = None


class GuardrailRuleResponse(ApiSchema):
    id: str
    chatbot_id: str
    rule_type: RuleType
    target_category: str | None = None
    match_mode: MatchMode
    match_value: str | None = None
    action_type: ActionType
    severity: Severity
    fallback_message: str | None = None
    escalation_message: str | None = None
    priority: int
    is_active: bool
    metadata_json: dict
    created_at: str
    updated_at: str


class GuardrailRuleListResponse(ApiSchema):
    rules: list[GuardrailRuleResponse]

File: apps/web/lib/api/guardrails.ts
import { apiClient } from "./client";

export type GuardrailRuleType =
  | "restricted_category"
  | "forbidden_phrase"
  | "escalation_trigger"
  | "sensitive_topic"
  | "response_constraint";

export type GuardrailActionType =
  | "restricted"
  | "escalate"
  | "ask_clarification"
  | "fallback"
  | "warn"
  | "require_cautious_wording";

export type GuardrailSeverity = "low" | "medium" | "high" | "critical";

export type GuardrailRule = {
  id: string;
  chatbotId: string;
  ruleType: GuardrailRuleType;
  targetCategory?: string | null;
  matchMode: string;
  matchValue?: string | null;
  actionType: GuardrailActionType;
  severity: GuardrailSeverity;
  priority: number;
  isActive: boolean;
  fallbackMessage?: string | null;
  escalationMessage?: string | null;
  metadataJson: Record<string, unknown>;
  createdAt: string;
  updatedAt: string;
};

export type GuardrailsResponse = {
  rules: GuardrailRule[];
};

export type GuardrailRuleCreateRequest = {
  ruleType: GuardrailRuleType;
  targetCategory?: string;
  matchMode?: "keyword_any" | "contains" | "exact" | "context_flag";
  matchValue?: string;
  actionType: GuardrailActionType;
  severity?: GuardrailSeverity;
  fallbackMessage?: string;
  escalationMessage?: string;
  priority?: number;
  isActive?: boolean;
  metadataJson?: Record<string, unknown>;
};

export type GuardrailRuleUpdateRequest = Partial<GuardrailRuleCreateRequest>;

export async function getGuardrails(chatbotId: string): Promise<GuardrailsResponse> {
  return apiClient.request<GuardrailsResponse>(`/admin/chatbots/${chatbotId}/guardrails`);
}

export async function createGuardrail(
  chatbotId: string,
  body: GuardrailRuleCreateRequest,
): Promise<GuardrailRule> {
  return apiClient.request<GuardrailRule>(`/admin/chatbots/${chatbotId}/guardrails`, {
    method: "POST",
    body,
  });
}

export async function patchGuardrail(
  chatbotId: string,
  ruleId: string,
  body: GuardrailRuleUpdateRequest,
): Promise<GuardrailRule> {
  return apiClient.request<GuardrailRule>(`/admin/chatbots/${chatbotId}/guardrails/${ruleId}`, {
    method: "PATCH",
    body,
  });
}

export async function deleteGuardrail(chatbotId: string, ruleId: string): Promise<void> {
  return apiClient.request<void>(`/admin/chatbots/${chatbotId}/guardrails/${ruleId}`, {
    method: "DELETE",
  });
}
